logic/tool/ExpressionResolution.java

```diff
@@ -0,0 +1,162 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.AMBIGUOUS_EXPRESSION_ARGUMENT_TYPE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.UNRESOLVABLE_FUNCTION;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Expression.UNRESOLVABLE_OPERATION;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Pattern.CYCLIC_VALUE_ASSIGNMENT;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Pattern.UNASSIGNED_VALUE_VARIABLE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class ExpressionResolution {
+    private final GraphManager graphMgr;
+
+    public ExpressionResolution(GraphManager graphMgr) {
+        this.graphMgr = graphMgr;
+    }
+
+    public void resolveExpressionTypes(Disjunction disjunction) {
+        List<ValueVariable> valueVariables = iterate(disjunction.conjunctions())
+                .flatMap(conjunction -> iterate(conjunction.variables()))
+                .filter(Variable::isValue).map(Variable::asValue).toList();
+
+        // TODO: Consider variables in nested patterns
+        iterate(valueVariables).filter(v -> v.assignedBy() == null)
+                .forEachRemaining(v -> {
+                    throw TypeDBException.of(UNASSIGNED_VALUE_VARIABLE, v);
+                });
+        Set<ValueVariable> safe = new HashSet<>();
+        Stack<ValueVariable> current = new Stack<>();
+        valueVariables.forEach(v -> validateCyclicAssignment(v, safe, current));
+
+        valueVariables.forEach(v -> resolveExpressionTypes(v.asValue()));
+    }
+
+    private void validateCyclicAssignment(ValueVariable at, Set<ValueVariable> safe, Stack<ValueVariable> current) {
+        if (safe.contains(at)) return;
+
+        if (current.contains(at)) {
+            current.push(at);
+            throw TypeDBException.of(CYCLIC_VALUE_ASSIGNMENT, current);
+        }
+
+        current.push(at);
+        iterate(at.assignedBy().inputs()).filter(Variable::isValue).forEachRemaining(v -> validateCyclicAssignment(v.asValue(), safe, current));
+        ValueVariable poppedAt = current.pop();
+        assert poppedAt == at;
+        safe.add(at);
+    }
+
+    private Encoding.ValueType<?> resolveExpressionTypes(ValueVariable v) {
+        if (v.assignedBy().returnType() == null) {
+            Map<Identifier, Encoding.ValueType<?>> inputTypes = new HashMap<>();
+            // No circular dependencies -> No infinite recursion
+            iterate(v.assignedBy().inputs()).forEachRemaining(input -> {
+                if (input.isThing()) {
+                    inputTypes.put(input.id(), ensureAndGetSingleValueType(input.asThing()));
+                } else if (v.isValue()) {
```

Comment:
It is indeed a bug.
There aren't any tests.
But they wouldn't have caught it either since the condition is always true (and is expected to be true).